### PR TITLE
Export Contents As Setup recipe

### PR DESCRIPTION
**Pull request by agriffard**
Add a setting in `Content `and `All Content` deployment steps to determine if the data have to be exported as a setup recipe.

### CODE REVIEW

**Comment by agriffard** [comment]
image

If you have a suggestion for a better property name instead of `ExportDatesAndIds` and the hint, please comment.

**Comment by Skrypt** [comment]
I'm not sure why you require this.

**Comment by agriffard** [comment]
When you want to export your data to a setup recipe.

It could be improved:
- Generate`[js: uuid()]` in ContentItemId.
- Generate`[js: parameters('AdminUsername')]` in Owner.
- If there is a List, put the id in a variable and use it into 
```
"ContainedPart": {
            "ListContentItemId": "[js: variables('listContentItemId')]",.
```

**Comment by Skrypt** [comment]
Ok then why not make the checkbox hint : Export as a setup recipe. And when it's unchecked then we assume that it is going to be generating a recipe.json file for actually make 2 tenants to get in sync.

**Comment by agriffard** [comment]
image

**Comment by agriffard** [comment]
Also added an `ExportDisableFeatures` setting in AllFeatures step.

**Comment by deanmarcussen** [comment]
Can you put a warning / hint on the text, to say that users will have to edit the recipe to supply the some of the variables / properties.

I see why you're wanting this, but it's dangerous, because if someone tries to run the output directly, it will cause problems with the list part / taxonomies / which are always refering to specific term content item ids, etc

**Comment by agriffard** [comment]
> Can you put a warning / hint on the text

Ok, done: 
image

File: src/OrchardCore.Modules/OrchardCore.Features/Deployment/AllFeaturesDeploymentSource.cs [hunk 1]
```diff
@@ -17,20 +17,24 @@ public AllFeaturesDeploymentSource(IModuleService moduleService)
...
-            result.Steps.Add(new JObject(
+            var featureStep = new JObject(
                 new JProperty("name", "Feature"),
-                new JProperty("enable", features.Where(f => f.IsEnabled).Select(f => f.Descriptor.Id).ToArray()),
-                new JProperty("disable", features.Where(f => !f.IsEnabled).Select(f => f.Descriptor.Id).ToArray())
-            ));
+                new JProperty("enable", features.Where(f => f.IsEnabled).Select(f => f.Descriptor.Id).ToArray())
+            );
+
+            if (allFeaturesStep.ExportDisableFeatures)
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Features/Deployment/AllFeaturesDeploymentSource.cs:33]: IgnoreDisabledFeature, left at false by default

File: src/OrchardCore.Modules/OrchardCore.Contents/Views/Items/AllContentDeploymentStep.Fields.Edit.cshtml [hunk 2]
```diff
@@ -1,3 +1,12 @@
...
 
 <h5>@T["All Content"]</h5>
+
+<div class="form-group">
+    <div class="custom-control custom-checkbox">
+        <input type="checkbox" class="custom-control-input" asp-for="ExportAsSetupRecipe" />
+        <label class="custom-control-label" asp-for="ExportAsSetupRecipe">@T["Export As Setup recipe"]</label>
+        <span class="hint">@T["Check if the data have to be exported as a Setup recipe."]</span>
+    </div>
+    <div class="alert alert-warning">@T["If you check this option, you will have to manually declare variables for the list container ids and taxonomies content ids and use them where needed in the recipe."]</div>
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Contents/Views/Items/AllContentDeploymentStep.Fields.Edit.cshtml:11]: Content Picker fields, Taxonomy fields, Tags fields, List content types.

Comment (sebastienros) [reply src/OrchardCore.Modules/OrchardCore.Contents/Views/Items/AllContentDeploymentStep.Fields.Edit.cshtml:11]: Any reference to these content items will be lost.

File: src/OrchardCore.Modules/OrchardCore.Contents/Views/Items/AllContentDeploymentStep.Fields.Edit.cshtml [hunk 3]
```diff
@@ -1,3 +1,12 @@
...
 
 <h5>@T["All Content"]</h5>
+
+<div class="form-group">
+    <div class="custom-control custom-checkbox">
+        <input type="checkbox" class="custom-control-input" asp-for="ExportAsSetupRecipe" />
+        <label class="custom-control-label" asp-for="ExportAsSetupRecipe">@T["Export As Setup recipe"]</label>
+        <span class="hint">@T["Check if the data should be exported as a Setup recipe."]</span>
+    </div>
+    <div class="alert alert-warning">@T["If checked, you will have to manually declare variables for the Content Picker fields, Taxonomy fields, Tags fields and List parts ids and replace them where needed in the recipe. Any reference to these content items will be lost."]</div>
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Contents/Views/Items/AllContentDeploymentStep.Fields.Edit.cshtml:11]:
```suggestion
    <div class="alert alert-warning">@T["If checked, you will have to manually declare variables for the Content Picker fields, Taxonomy fields, Tags fields and List item ids and replace them where needed in the recipe. Any reference to these content items will be lost."]</div>
```

File: src/OrchardCore.Modules/OrchardCore.Contents/Views/Items/ContentDeploymentStep.Fields.Edit.cshtml [hunk 4]
```diff
@@ -8,6 +8,15 @@
 <h5>@T["Content Types"]</h5>
 <span class="hint">@T["The content types to add as part of the plan."]</span>
 
+<div class="form-group">
+    <div class="custom-control custom-checkbox">
+        <input type="checkbox" class="custom-control-input" asp-for="ExportAsSetupRecipe" />
+        <label class="custom-control-label" asp-for="ExportAsSetupRecipe">@T["Export As Setup recipe"]</label>
+        <span class="hint">@T["Check if the data should be exported as a Setup recipe."]</span>
+    </div>
+    <div class="alert alert-warning">@T["If checked, you will have to manually declare variables for the Content Picker fields, Taxonomy fields, Tags fields and List parts ids and replace them where needed in the recipe. Any reference to these content items will be lost."]</div>
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Contents/Views/Items/ContentDeploymentStep.Fields.Edit.cshtml:17]:
```suggestion
    <div class="alert alert-warning">@T["If checked, you will have to manually declare variables for the Content Picker fields, Taxonomy fields, Tags fields and List item ids and replace them where needed in the recipe. Any reference to these content items will be lost."]</div>
```

File: src/OrchardCore.Modules/OrchardCore.Features/Deployment/AllFeaturesDeploymentSource.cs [hunk 5]
```diff
@@ -25,12 +25,16 @@ public async Task ProcessDeploymentStepAsync(DeploymentStep step, DeploymentPlan
...
-            result.Steps.Add(new JObject(
+            var featureStep = new JObject(
                 new JProperty("name", "Feature"),
-                new JProperty("enable", features.Where(f => f.IsEnabled).Select(f => f.Descriptor.Id).ToArray()),
-                new JProperty("disable", features.Where(f => !f.IsEnabled).Select(f => f.Descriptor.Id).ToArray())
-            ));
+                new JProperty("enable", features.Where(f => f.IsEnabled).Select(f => f.Descriptor.Id).ToArray())
+            );
+
+            if (!allFeaturesStep.IgnoreDisabledFeature)
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Features/Deployment/AllFeaturesDeploymentSource.cs:33]: Suggest `IgnoreDisabledFeatures`

File: src/OrchardCore.Modules/OrchardCore.Contents/Views/Items/AllContentDeploymentStep.Fields.Edit.cshtml (hunk 3 above)
Comment (agriffard) [reply src/OrchardCore.Modules/OrchardCore.Contents/Views/Items/AllContentDeploymentStep.Fields.Edit.cshtml]: ok

File: src/OrchardCore.Modules/OrchardCore.Contents/Views/Items/ContentDeploymentStep.Fields.Edit.cshtml (hunk 4 above)
Comment (agriffard) [reply src/OrchardCore.Modules/OrchardCore.Contents/Views/Items/ContentDeploymentStep.Fields.Edit.cshtml:17]: ok

File: src/OrchardCore.Modules/OrchardCore.Features/Deployment/AllFeaturesDeploymentSource.cs (hunk 5 above)
Comment (agriffard) [reply src/OrchardCore.Modules/OrchardCore.Features/Deployment/AllFeaturesDeploymentSource.cs]: ok IgnoreDisabledFeatures

**Pull request merged**